### PR TITLE
Add UserGroup & membership queries

### PR DESCRIPTION
**Pull request by priyesh777**
## Depends:
- https://github.com/the-deep/server/pull/1054

## Addresses:
- https://github.com/the-deep/client/issues/2180#issue-1017836853

## Changes:
- Replace the Rest APIs with the GQL queries for user-group elements and membership elements

## This PR doesn't introduce any:

- [x] temporary files, auto-generated files or secret keys
- [x] build works
- [ ] eslint issues
- [x] typescript issues
- [x] codegen errors
- [x] `console.log` meant for debugging
- [x] typos
- [x] unwanted comments
- [x] conflict markers

## This PR contains valid:

- [ ] permission checks
- [ ] translations

### CODE REVIEW

File: app/types/project.tsx [hunk 1]
```diff
@@ -148,20 +148,56 @@ export interface ProjectsSummary {
...
+//    usergroup: number;
+//    role: number;
+//    addedBy: number;
+//    addedByName: string;
+//    roleDetails: {
+//        id: number;
+//        title: string;
+//        level: number;
+//    };
+// }
```
Comment (samshara) [inline app/types/project.tsx]: let's remove this commented code.

File: app/views/Project/index.tsx [hunk 2]
```diff
@@ -69,13 +69,13 @@ function Project(props: Props) {
...
             onCompleted: (data) => {
-                if (data.project && data.project.allowedPermissions.length > 0) {
+                if (data.project && data.project.allowedPermissions?.length > 0) {
                     setProject(data.project);
 
                     setLastActiveProject({
                         variables: { id: data.project.id },
                     });
-                }
+                } return null;
```
Comment (samshara) [inline app/views/Project/index.tsx]: do we need to return `null` here

File: app/views/ProjectEdit/Users/UserGroupList/AddUserGroupModal/index.tsx [hunk 3]
```diff
@@ -31,25 +31,33 @@ import NonFieldError from '#components/NonFieldError';
...
+    id: string;
+    usergroup: {
+        id: string;
+        title: string;
+    }
 }
 
 const usergroupKeySelector = (d: UserGroupMini) => d.id;
-const usergroupLabelSelector = (d: UserGroupMini) => d.title;
+const usergroupLabelSelector = (d: UserGroupMini) => d.usergroup.title;
```
Comment (samshara) [inline app/views/ProjectEdit/Users/UserGroupList/AddUserGroupModal/index.tsx]: we are using `usergroup.title` for label and `id` for id. There is `usergroup.id` field too. Please check which id is the valid one.

File: app/views/ProjectEdit/Users/index.tsx [hunk 4]
```diff
@@ -21,21 +50,25 @@ function Users(props: Props) {
         activeUserId,
```
Comment (tnagorra) [inline app/views/ProjectEdit/Users/index.tsx:51]: Do we not need activeUserId anymore?

File: app/views/ProjectEdit/Users/index.tsx [hunk 5]
```diff
@@ -1,15 +1,44 @@
...
+} from '#generated/types';
 
 import styles from './styles.css';
 
+const MEMBERSHIP = gql`
+    query Membership($id: ID!) {
+        project(id: $id) {
+            userMembers {
+                results {
+                  clientId
```
Comment (tnagorra) [inline app/views/ProjectEdit/Users/index.tsx:19]: Do we need to get clientId?

File: app/views/ProjectEdit/Users/UserGroupList/index.tsx [hunk 6]
```diff
@@ -71,25 +121,29 @@ function UserGroupList(props: Props) {
         setModalHidden();
     }, [setModalHidden]);
 
-    const queryForRequest = useMemo(() => ({
-        offset: (activePage - 1) * maxItemsPerPage,
-        limit: maxItemsPerPage,
-    }), [activePage]);
+    const userGroupVariables = useMemo(
+        (): UserGroupMembersQueryVariables | undefined => ({
```
Comment (tnagorra) [inline app/views/ProjectEdit/Users/UserGroupList/index.tsx:123]:
```suggestion
        (): UserGroupMembersQueryVariables => ({
```

File: app/views/ProjectEdit/Users/UserGroupList/AddUserGroupModal/index.tsx [hunk 7]
```diff
@@ -31,25 +31,33 @@ import NonFieldError from '#components/NonFieldError';
...
 const roleLabelSelector = (d: ProjectRole) => d.title;
 
 type FormType = {
-    id?: number;
-    usergroup?: number;
+    id?: string;
+    usergroup: {
+        id: string;
+        title: string;
+    }
```
Comment (tnagorra) [inline app/views/ProjectEdit/Users/UserGroupList/AddUserGroupModal/index.tsx]: Form should not include usergroup as object.

File: app/types/project.tsx [hunk 8]
```diff
@@ -148,20 +148,56 @@ export interface ProjectsSummary {
...
+export interface UserGroupMemberships {
+    id: string;
+    joinedAt: string;
+    member: {
+        id: string;
+        displayName?: string | null | undefined;
+    };
+}
+
 export interface UserGroup {
```
Comment (tnagorra) [inline app/types/project.tsx:162]: Any reason we are modfying this Type instead of removing this

**Review by tnagorra — CHANGES_REQUESTED** [review]

File: app/views/ProjectEdit/Users/index.tsx (hunk 4 above)
Comment (priyesh777) [reply app/views/ProjectEdit/Users/index.tsx:51]: Yes it has been used to disable a particular column in a table conditionally.

File: app/types/project.tsx (hunk 8 above)
Comment (priyesh777) [reply app/types/project.tsx:162]: The new type changes had been included in that particular interface due which it couldn't be removed.

File: app/views/ProjectEdit/Users/index.tsx (hunk 4 above)
Comment (tnagorra) [reply app/views/ProjectEdit/Users/index.tsx:51]: We should discuss further about this requirement

File: app/views/ProjectEdit/Users/UserGroupList/AddUserGroupModal/index.tsx [hunk 9]
```diff
@@ -120,14 +152,22 @@ function AddUserGroupModal(props: Props) {
...
-        query: queryForUsergroups,
-    });
+        data: userGroupsResponse,
+        loading: pendingUserGroups,
+    } = useQuery<UserGroupsQuery, UserGroupsQueryVariables>(
+        USERGROUPS,
+        {
+            variables: membershipVariables,
+        },
+    );
```
Comment (AdityaKhatri) [inline app/views/ProjectEdit/Users/UserGroupList/AddUserGroupModal/index.tsx:170]: Create usergroup search multiselect input

File: app/views/ProjectEdit/Users/UserGroupList/AddUserGroupModal/index.tsx [hunk 10]
```diff
@@ -216,7 +243,7 @@ function AddUserGroupModal(props: Props) {
             <NonFieldError error={error} />
             <SelectInput
                 name="usergroup"
-                options={usergroupList}
+                options={userGroupsResponse?.userGroups?.results}
```
Comment (AdityaKhatri) [inline app/views/ProjectEdit/Users/UserGroupList/AddUserGroupModal/index.tsx:246]: the default options should be provided in multiselect search input

**Review by AdityaKhatri — CHANGES_REQUESTED** [review]

**Comment by AdityaKhatri** [comment]
Closed because of this https://github.com/the-deep/client/pull/2375

**Pull request closed without merging**